### PR TITLE
Half life for auto compounding rewards

### PR DESCRIPTION
**Pull request by barakman**

### CODE REVIEW

File: contracts/rewards/AutoCompoundingRewards.sol [hunk 1]
```diff
@@ -33,6 +33,8 @@ import {
 
 import { RewardsMath } from "./RewardsMath.sol";
 
+uint32 constant SECONDS_PER_DAY = 60 * 60 * 24;
```
Comment (yudilevi) [inline contracts/rewards/AutoCompoundingRewards.sol:36]: I believe you can just use `1 day` here or just get rid of it and do the multiplication later by `1 day`

Comment (barakman) [reply contracts/rewards/AutoCompoundingRewards.sol:36]: Yes I can

File: contracts/rewards/AutoCompoundingRewards.sol [hunk 2]
```diff
@@ -212,7 +213,8 @@ contract AutoCompoundingRewards is IAutoCompoundingRewards, ReentrancyGuardUpgra
         uint256 totalRewards,
         uint8 distributionType,
         uint32 startTime,
-        uint32 endTime
+        uint32 endTime,
+        uint32 halfLifeInDays
```
Comment (lbeder) [inline contracts/rewards/AutoCompoundingRewards.sol:217]: Since it's also irrelevant for flat distribution, I'd suggest stopping using a generic single `createProgram` function and splitting it into two specific functions with their own parameters and validation (and sharing the common logic via another `private` function)

File: contracts/rewards/RewardsMath.sol [hunk 3]
```diff
@@ -25,14 +22,18 @@ library RewardsMath {
 
     /**
      * @dev returns the amount of rewards distributed after a given time period since deployment has elapsed
-     * The returned value is calculated as `totalRewards * (1 - 1 / e ^ (timeElapsed * LAMBDA))`.
-     * Note that because the exponentiation function is limited to an input of up to (and excluding) 16, the
-     * input value to this function is limited by `timeElapsed * LAMBDA < 16` --> `timeElapsed < 1120000000`.
-     * For `timeElapsed = 1120000000 - 1`, the formula above returns more than 99.9999% of `totalRewards`.
+     *
+     * The returned value is calculated as `totalRewards * (1 - 1 / 2 ^ (timeElapsed / halfLife))`
```
Comment (lbeder) [inline contracts/rewards/RewardsMath.sol:26]: (Comments should start with lowercase letters)

File: test/rewards/AutoCompoundingRewards.ts [hunk 4]
```diff
@@ -189,11 +189,22 @@ describe('AutoCompoundingRewards', () => {
...
-        const TOTAL_DURATION = duration.days(10);
         const TOTAL_REWARDS = toWei(10_000);
         const INITIAL_USER_STAKE = toWei(10_000);
 
         const START_TIME = 1000;
+        const FLAT_TOTAL_DURATION = duration.days(10);
+        const EXP_DECAY_HALF_LIFE = duration.days(560);
+        const EXP_DECAY_MAX_DURATION = EXP2_INPUT_TOO_HIGH.mul(EXP_DECAY_HALF_LIFE).sub(1).ceil().toNumber();
+
+        const END_TIME = (distributionType: RewardsDistributionType) =>
```
Comment (lbeder) [inline test/rewards/AutoCompoundingRewards.ts:200]: Since these are functions, they shouldn't be named as `CONSTANTS`

**Review by lbeder — CHANGES_REQUESTED** [review]

File: contracts/rewards/AutoCompoundingRewards.sol (hunk 2 above)
Comment (barakman) [reply contracts/rewards/AutoCompoundingRewards.sol]: done

File: test/rewards/AutoCompoundingRewards.ts (hunk 4 above)
Comment (barakman) [reply test/rewards/AutoCompoundingRewards.ts]: done

File: contracts/rewards/RewardsMath.sol (hunk 3 above)
Comment (barakman) [reply contracts/rewards/RewardsMath.sol]: done

File: contracts/rewards/AutoCompoundingRewards.sol (hunk 1 above)
Comment (barakman) [reply contracts/rewards/AutoCompoundingRewards.sol]: done

File: contracts/rewards/AutoCompoundingRewards.sol [hunk 5]
```diff
@@ -206,16 +206,16 @@ contract AutoCompoundingRewards is IAutoCompoundingRewards, ReentrancyGuardUpgra
     }
 
     /**
-     * @inheritdoc IAutoCompoundingRewards
+     * @dev creates a program for a pool
      */
-    function createProgram(
+    function _createProgram(
```
Comment (lbeder) [inline contracts/rewards/AutoCompoundingRewards.sol:211]: Please move private/internal function after all external/public functions

File: test/rewards/AutoCompoundingRewards.ts [hunk 6]
```diff
@@ -197,15 +197,32 @@ describe('AutoCompoundingRewards', () => {
...
-        const HALF_LIFE = (distributionType: RewardsDistributionType) =>
-            distributionType === RewardsDistributionType.Flat ? 0 : EXP_DECAY_HALF_LIFE / duration.days(1);
+        const programHalfLife = (distributionType: RewardsDistributionType) =>
+            distributionType === RewardsDistributionType.Flat ? 0 : EXP_DECAY_HALF_LIFE;
 
-        const DURATION = (distributionType: RewardsDistributionType) =>
+        const programDuration = (distributionType: RewardsDistributionType) =>
             distributionType === RewardsDistributionType.Flat ? FLAT_TOTAL_DURATION : EXP_DECAY_MAX_DURATION;
 
+        const createProgram = async (
```
Comment (lbeder) [inline test/rewards/AutoCompoundingRewards.ts:209]: I'd add `createFlatProgram` and `createExpProgram` here as well so that the tests/profiling is clearer as well

File: test/rewards/AutoCompoundingRewards.ts [hunk 7]
```diff
@@ -197,15 +197,32 @@ describe('AutoCompoundingRewards', () => {
         const EXP_DECAY_HALF_LIFE = duration.days(560);
         const EXP_DECAY_MAX_DURATION = EXP2_INPUT_TOO_HIGH.mul(EXP_DECAY_HALF_LIFE).sub(1).ceil().toNumber();
 
-        const END_TIME = (distributionType: RewardsDistributionType) =>
+        const programEndTime = (distributionType: RewardsDistributionType) =>
             distributionType === RewardsDistributionType.Flat ? START_TIME + FLAT_TOTAL_DURATION : 0;
 
-        const HALF_LIFE = (distributionType: RewardsDistributionType) =>
-            distributionType === RewardsDistributionType.Flat ? 0 : EXP_DECAY_HALF_LIFE / duration.days(1);
+        const programHalfLife = (distributionType: RewardsDistributionType) =>
```
Comment (lbeder) [inline test/rewards/AutoCompoundingRewards.ts:203]: These functions just need to be hashes

**Review by lbeder — CHANGES_REQUESTED** [review]

File: contracts/rewards/AutoCompoundingRewards.sol (hunk 5 above)
Comment (barakman) [reply contracts/rewards/AutoCompoundingRewards.sol:211]: Sure, but the purpose was to improve the quality of the review by reducing the number of changes to a minimum.
@yudilevi : Please inform when you have completed reviewing this part.

File: test/rewards/AutoCompoundingRewards.ts (hunk 6 above)
Comment (barakman) [reply test/rewards/AutoCompoundingRewards.ts:209]: Are you sure?
That would "mess up" (increase in size) all the tests which use this function.
The idea was to keep them "clean" and handle this "duality" in a single place.

File: test/rewards/AutoCompoundingRewards.ts (hunk 7 above)
Comment (barakman) [reply test/rewards/AutoCompoundingRewards.ts:203]: What does that mean?

File: contracts/rewards/AutoCompoundingRewards.sol (hunk 5 above)
Comment (yudilevi) [reply contracts/rewards/AutoCompoundingRewards.sol:211]: you can move it, it's not an issue for the review

File: contracts/rewards/AutoCompoundingRewards.sol [hunk 8]
```diff
@@ -264,10 +258,43 @@ contract AutoCompoundingRewards is IAutoCompoundingRewards, ReentrancyGuardUpgra
...
+            revert InvalidParam();
+        }
+
+        _createProgram(pool, totalRewards, FLAT_DISTRIBUTION, startTime, endTime, 0);
+    }
+
+    /**
+     * @inheritdoc IAutoCompoundingRewards
+     */
+    function createExpProgram(
```
Comment (yudilevi) [inline contracts/rewards/AutoCompoundingRewards.sol:285]: It's ExpDecay

File: contracts/rewards/interfaces/IAutoCompoundingRewards.sol [hunk 9]
```diff
@@ -44,21 +45,35 @@ interface IAutoCompoundingRewards is IUpgradeable {
     function isProgramActive(Token pool) external view returns (bool);
 
     /**
-     * @dev creates a program for a pool
+     * @dev creates a flat program for a pool
```
Comment (yudilevi) [inline contracts/rewards/interfaces/IAutoCompoundingRewards.sol:48]: thoughts re.  changing to `creates a rewards program for a pool, with flat distribution`?

File: test/rewards/AutoCompoundingRewards.ts (hunk 6 above)
Comment (lbeder) [reply test/rewards/AutoCompoundingRewards.ts:209]: It's a good tradeoff. Although it'll increase the code, it'll be always 100% clear how each program is set.

File: test/rewards/AutoCompoundingRewards.ts (hunk 7 above)
Comment (lbeder) [reply test/rewards/AutoCompoundingRewards.ts:203]: Something like this:

```
 const programHalfLives = {
    [RewardsDistributionType.Flat]: 0,
    [RewardsDistributionType.ExponentialDecay]: EXP_DECAY_HALF_LIFE
};
 ```

File: test/rewards/AutoCompoundingRewards.ts (hunk 6 above)
Comment (barakman) [reply test/rewards/AutoCompoundingRewards.ts:209]: But it'll just move the `distributionType === RewardsDistributionType.Flat ?` check from this function into each and every one of the tests which use it.

Comment (lbeder) [reply test/rewards/AutoCompoundingRewards.ts:209]: It will also abstract away distribution-specific inputs (like in the contracts themselves), I believe

File: test/rewards/AutoCompoundingRewards.ts (hunk 7 above)
Comment (barakman) [reply test/rewards/AutoCompoundingRewards.ts]: That's a really good idea!
Done.

File: contracts/rewards/AutoCompoundingRewards.sol (hunk 5 above)
Comment (barakman) [reply contracts/rewards/AutoCompoundingRewards.sol]: Done

File: contracts/rewards/AutoCompoundingRewards.sol (hunk 8 above)
Comment (barakman) [reply contracts/rewards/AutoCompoundingRewards.sol]: Done (as part of a conclusive rename of 'ExponentialDecay' to 'ExpDecay').

File: contracts/rewards/interfaces/IAutoCompoundingRewards.sol (hunk 9 above)
Comment (barakman) [reply contracts/rewards/interfaces/IAutoCompoundingRewards.sol]: Done:
-  creates a rewards program with flat distribution for a given pool
-  creates a rewards program with exponential-decay distribution for a given pool

File: test/rewards/AutoCompoundingRewards.ts (hunk 6 above)
Comment (yudilevi) [reply test/rewards/AutoCompoundingRewards.ts:215]: can also just separate the tests and not using `if` in most cases (although I can see how that will be annoying with general management functions).
another approach would be to create separate functions + another function that receives a `type` arg to be used in the management tests, while calling the dedicated functions directly in the financial tests.
but in general I agree that having a function with all permutations is less readable.

Comment (lbeder) [reply test/rewards/AutoCompoundingRewards.ts:215]: The latter is how I imagined it as well

File: contracts/rewards/AutoCompoundingRewards.sol [hunk 10]
```diff
@@ -383,7 +383,7 @@ contract AutoCompoundingRewards is IAutoCompoundingRewards, ReentrancyGuardUpgra
             return
                 RewardsMath.calcFlatRewards(p.totalRewards, currTimeElapsed - prevTimeElapsed, p.endTime - p.startTime);
         } else {
-            // if (p.distributionType == EXPONENTIAL_DECAY_DISTRIBUTION)
+            // if (p.distributionType == EXP_DECAY_DISTRIBUTION)
```
Comment (lbeder) [inline contracts/rewards/AutoCompoundingRewards.sol:386]: Do we really need this comment?

**Review by lbeder — DISMISSED** [review]

File: contracts/rewards/AutoCompoundingRewards.sol (hunk 10 above)
Comment (barakman) [reply contracts/rewards/AutoCompoundingRewards.sol:386]: IMO, it covers for the inherent lack of a (self-explanatory) switch/case statement.

Comment (lbeder) [reply contracts/rewards/AutoCompoundingRewards.sol:386]: It looks like an accidental copy-paste from the contracts. Can you rephrase it?

File: test/rewards/AutoCompoundingRewards.ts (hunk 6 above)
Comment (barakman) [reply test/rewards/AutoCompoundingRewards.ts:215]: done

File: contracts/rewards/AutoCompoundingRewards.sol (hunk 10 above)
Comment (barakman) [reply contracts/rewards/AutoCompoundingRewards.sol]: done

**Review by yudilevi — APPROVED** [review]

**Review by lbeder — APPROVED** [review]

**Review by yudilevi — APPROVED** [review]
